test: add unit tests for restore.ts file

rafaquelhodev opened the pull request:
Create unit tests for `restore.ts` file. fixes #3307

ad1992 reviewed (changes requested):
Thanks for working on this @rafaquelhodev 
can you break the specs into multiple describe blocks, one describe per function similar to how we have it in https://github.com/excalidraw/excalidraw/blob/master/src/tests/packages/utils.test.ts
Also, the file can be moved to `src/tests/data/restore.test.ts` since it's pure utility so it can bu simply `ts` file.

rafaquelhodev commented:
No worries @ad1992 , this is a cool project :)

I've just improved the describe blocks as you mentioned.

I created a `.tsx` file because I had to use the `UI` to create a `selection` element (`UI.createElement("selection")`). Therefore, I had to insert a 
`
beforeEach(async () => {
  await render(<ExcalidrawApp/>)
})
` in the tests.

If you want I can try to create a selection element in a different manner, so we don't have to `render(<ExcalidrawApp/>)` before each test.

ad1992 commented:
@rafaquelhodev we don't need to render `<ExcalidrawApp/>` since we are unit testing we just need to test the functions in `restore.ts` as pure utilities as we did in https://github.com/excalidraw/excalidraw/blob/master/src/tests/packages/utils.test.ts , passing the arguments and asserting for various scenarios. 
Let me know if you have any questions

rafaquelhodev commented:
Hi @ad1992 , the point is that I created a test to test the behaviour of `restoreElements` function when we have a `selection` element type. But I was not able to create this element with the `API.createElement({ type: "selection" })` function. Do you have any suggestions on how to test the `restoreElements` function for a `selection` element type?

rafaquelhodev commented:
Hi @ad1992 , I realized just now what you were saying, my bad. I've corrected the tests and moved the file to `src/tests/data/restore.test.ts`. Now we don't have any dependency to `<ExcalidrawApp/>` =)

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+});
+
+describe("restoreElements", () => {
+  it("with null array of elements", () => {
```
```suggestion
  it("should return empty array when element is null", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    expect(restore.restoreElements(null)).toStrictEqual([]);
+  });
+
+  it("should not call isInvisiblySmallElement when input element is a selection element", () => {
```
```suggestion
  it("should not call isInvisiblySmallElement when element is a selection element", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    expect(sizeHelpers.isInvisiblySmallElement).toBeCalledTimes(0);
+  });
+
+  it("when input element type is not supported", () => {
```
```suggestion
  it("should return empty array when input type is not supported", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    expect(restore.restoreElements([dummyNotSupportedElement]).length).toBe(0);
+  });
+
+  it("when isInvisiblySmallElement is true", () => {
```
```suggestion
  it("should return empty array when isInvisiblySmallElement is true", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+      type: "text",
+      fontSize: 14,
+      fontFamily: 1,
+      text: "text",
+      textAlign: "center",
+      verticalAlign: "middle",
```
```suggestion
```

let's remove this so restoring correctly is asserted

rafaquelhodev replied:
as we talked before, I've separated the tests for text elements in two: 1) passing all attributes 2) not passing all attributes and letting the restore function do its job :)

rafaquelhodev replied:
@ad1992 I updated the tests with the snapshots :)

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    expect(restore.restoreElements([rectElement]).length).toBe(0);
+  });
+
+  it("with text element type", () => {
```
```suggestion
  it("should restore text element correctly", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+      verticalAlign: "middle",
+    });
+
+    const expectedTextElement = {
+      type: textElement.type,
+      fontSize: textElement.fontSize,
+      fontFamily: textElement.fontFamily,
+      text: textElement.text,
+      baseline: textElement.baseline,
+      textAlign: textElement.textAlign,
+      verticalAlign: textElement.verticalAlign,
+    };
+
+    const restoredText = restore.restoreElements([
+      textElement,
+    ])[0] as ExcalidrawTextElement;
+
+    const gotRestoredTextElement = {
+      type: restoredText.type,
+      fontSize: restoredText.fontSize,
+      fontFamily: restoredText.fontFamily,
+      text: restoredText.text,
+      baseline: restoredText.baseline,
+      textAlign: restoredText.textAlign,
+      verticalAlign: restoredText.verticalAlign,
+    };
+
+    expect(gotRestoredTextElement).toMatchObject(expectedTextElement);
```
lets take a snapshot instead, will be easier as there are many attributes

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    expect(gotRestoredTextElement).toMatchObject(expectedTextElement);
+  });
+
+  it("when text element has null text", () => {
```
```suggestion
  it("should set text to empty string when text element text is null", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    expect(restoredText.text).toBe("");
+  });
+
+  it("when text element has undefined alignment", () => {
```
instead of having separate assertions for all conditons, lets have two snapshots
1. One where you pass value for each attribute of text element which you are already doing above
1. One as suggested above where we can remove the fields which are set to default values

rafaquelhodev replied:
Hi @ad1992 ! Thanks for your suggestion, indeed we need to snapshot the tests! I was wondering if we can have 3 tests for the text elements:
```javascript
// test 01
// passing value for each attribute
const textElement = API.createElement({
type:  "text",
fontSize:  14,
fontFamily:  1,
text:  "text",
textAlign:  "center",
verticalAlign:  "middle"
});
expect(restoredText).toMatchSnapshot({seed: expect.any(Number)});
```

```javascript
// test 02
// font property, text null, alignment undefined
const textElement = API.createElement({
type:  "text",
text: null,
font: "10 Cascadia",
textAlign: undefined,
verticalAlign: undefined
});
expect(restoredText).toMatchSnapshot({seed: expect.any(Number)});
```

```javascript
// test 03
// font property with unknown font family, text null, alignment undefined
const textElement = API.createElement({
type:  "text",
text: null,
font: "10 unknown",
textAlign: undefined,
verticalAlign: undefined
});
expect(restoredText).toMatchSnapshot({seed: expect.any(Number)});
```

What do you think?

ad1992 replied:
in third case only font family is unknown so lets merge it with the second case?

rafaquelhodev replied:
sounds good to me!

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    const restoredLine = restoredElements[0] as ExcalidrawLinearElement;
+    const restoredDraw = restoredElements[1] as ExcalidrawLinearElement;
+
+    const gotRestoredLineElement = {
+      type: restoredLine.type,
+      startBinding: restoredLine.startBinding,
+      endBinding: restoredLine.endBinding,
+      lastCommittedPoint: restoredLine.lastCommittedPoint,
+      startArrowhead: restoredLine.startArrowhead,
+      endArrowhead: restoredLine.endArrowhead,
+    };
+
+    const gotRestoredDrawElement = {
+      type: restoredDraw.type,
+      startBinding: restoredDraw.startBinding,
+      endBinding: restoredDraw.endBinding,
+      lastCommittedPoint: restoredDraw.lastCommittedPoint,
+      startArrowhead: restoredDraw.startArrowhead,
+      endArrowhead: restoredDraw.endArrowhead,
+    };
+
+    expect(gotRestoredLineElement).toMatchObject(expectedRestoredLineElement);
+    expect(gotRestoredDrawElement).toMatchObject(expectedRestoredDrawElement);
```
lets have snapshot

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+      startArrowhead: restoredArrow.startArrowhead,
+    };
+
+    expect(gotRestoredArrowElement).toMatchObject(expectedRestoredArrowElement);
```
snapshot here as well

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+    };
+
+    expect(restoredRect.x).toBe(10);
+    expect(restoredRect).toMatchObject(expectedRestoredRect);
```
snapshot

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,703 @@
+
+    const restoredElements = restore.restoreElements(elements);
+
+    expect(restoredElements[0]).toMatchObject(expectedRestoredRect);
+    expect(restoredElements[1]).toMatchObject(expectedRestoredEllipse);
+    expect(restoredElements[2]).toMatchObject(expectedRestoredDiamond);
```
snapshot

ad1992 commented:
> Hi @ad1992 , I realized just now what you were saying, my bad. I've corrected the tests and moved the file to `src/tests/data/restore.test.ts`. Now we don't have any dependency to `<ExcalidrawApp/>` =)

no worries :)

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,527 @@
+      id: "id-text01",
+    });
+
+    Object.defineProperty(textElement, "text", {
+      get: jest.fn(() => null),
+    });
+
+    Object.defineProperty(textElement, "font", {
+      get: jest.fn(() => "10 unknown"),
+    });
```
you can pass these fields to `API.createElement` ?

rafaquelhodev replied:
@ad1992 Unfortunately, the `API.CreateElement` does not accept the property `text` with `null` value. Also, it does not accept the property `font`. Therefore, I had to use jest to insert these parameters.

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,527 @@
+    });
+  });
+
+  it("with freedraw element", () => {
```
```suggestion
  it("should restore freedraw element correctly", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,527 @@
+      pressures: restoredFreedraw.pressures,
+    };
+
+    expect(gotRestoredFreedrawElement).toMatchObject(expectedFreedrawElement);
```
lets have snapshot

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,527 @@
+    expect(gotRestoredFreedrawElement).toMatchObject(expectedFreedrawElement);
+  });
+
+  it("with line and draw elements", () => {
```
```suggestion
  it("should restore line and draw elements correctly", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,527 @@
+    expect(restoredDraw).toMatchSnapshot({ seed: expect.any(Number) });
+  });
+
+  it("with arrow element", () => {
```
```suggestion
  it("should restore arrow element correctly", () => {
```

ad1992 commented on src/tests/data/restore.test.ts:
```diff
@@ -0,0 +1,527 @@
+  });
+
+  it("should return current app state when imported data state is undefined", () => {
+    const stubImportedAppState = getDefaultAppState();
+
+    Object.defineProperty(stubImportedAppState, "cursorButton", {
+      get: jest.fn(() => undefined),
+    });
+
+    Object.defineProperty(stubImportedAppState, "name", {
+      get: jest.fn(() => undefined),
+    });
```
```suggestion
    const stubImportedAppState =  {...getDefaultAppState(), cursorButton: undefined, name: undefined };

```

ad1992 commented on src/tests/helpers/api.ts:
```diff
@@ -157,6 +157,7 @@ export class API {
         });
         break;
     }
+
```
```suggestion
```

ad1992 reviewed (approved):
Thats a lot of specs 🎉 , thanks @rafaquelhodev 💯

rafaquelhodev commented:
thanks for the review @ad1992 :)
